image/tarexport/save.go

```diff
@@ -608,22 +616,82 @@ func (s *saveSession) saveConfig(legacyImg image.V1Image, outDir string, created
 		return err
 	}
 
+	ts := time.Unix(0, 0)
+	if createdTime != nil {
+		ts = *createdTime
+	}
+
 	cfgDgst := digest.FromBytes(imageConfig)
 	configPath := filepath.Join(outDir, cfgDgst.Algorithm().String(), cfgDgst.Encoded())
-	if err := os.MkdirAll(filepath.Dir(configPath), 0o755); err != nil {
+	if err := mkdirAllWithChtimes(filepath.Dir(configPath), 0o755, ts, ts); err != nil {
 		return errors.Wrap(err, "could not create layer dir parent")
 	}
 
 	if err := os.WriteFile(configPath, imageConfig, 0o644); err != nil {
 		return err
 	}
 
-	if createdTime != nil {
-		if err := system.Chtimes(configPath, *createdTime, *createdTime); err != nil {
-			return errors.Wrap(err, "could not set config timestamp")
-		}
+	if err := system.Chtimes(configPath, ts, ts); err != nil {
+		return errors.Wrap(err, "could not set config timestamp")
 	}
 
 	s.savedConfigs[legacyImg.ID] = struct{}{}
 	return nil
 }
+
+// mkdirAllWithChtimes is nearly an identical copy to the os.MkdirAll but
+// tracks created directories and applies the provided mtime and atime using
+// system.Chtimes.
+func mkdirAllWithChtimes(path string, perm os.FileMode, atime, mtime time.Time) error {
+	// Fast path: if we can tell whether path is a directory or file, stop with success or error.
+	dir, err := os.Stat(path)
+	if err == nil {
+		if dir.IsDir() {
+			return nil
+		}
+		return &os.PathError{Op: "mkdir", Path: path, Err: syscall.ENOTDIR}
+	}
+
+	// Slow path: make sure parent exists and then call Mkdir for path.
+
+	// Extract the parent folder from path by first removing any trailing
+	// path separator and then scanning backward until finding a path
+	// separator or reaching the beginning of the string.
+	i := len(path) - 1
+	for i >= 0 && os.IsPathSeparator(path[i]) {
+		i--
+	}
+	for i >= 0 && !os.IsPathSeparator(path[i]) {
+		i--
+	}
+	if i < 0 {
+		i = 0
+	}
+
+	// If there is a parent directory, and it is not the volume name,
+	// recurse to ensure parent directory exists.
+	if parent := path[:i]; len(parent) > len(filepath.VolumeName(path)) {
+		err = mkdirAllWithChtimes(parent, perm, atime, mtime)
+		if err != nil {
+			return err
+		}
+	}
+
+	// Parent now exists; invoke Mkdir and use its result.
+	err = os.Mkdir(path, perm)
+	if err != nil {
+		// Handle arguments like "foo/." by
+		// double-checking that directory doesn't exist.
+		dir, err1 := os.Lstat(path)
+		if err1 == nil && dir.IsDir() {
+			return nil
+		}
+		return err
+	}
+
+	if err := system.Chtimes(path, atime, mtime); err != nil {
+		return fmt.Errorf("applying atime=%v and mtime=%v to %s: %w",
+			atime, mtime, path, err)
```

Comment:
Just noticed that the `to <Path>` may be redundant here, because `system.Chtimes` uses `os.Chtimes`, which returns a `PathError` already; 

https://github.com/golang/go/blob/39f2032c17516a55c165d329dd5e2e07f49132b0/src/os/file_posix.go#L190-L192

```go
if e := syscall.UtimesNano(fixLongPath(name), utimes[0:]); e != nil {
	return &PathError{Op: "chtimes", Path: name, Err: e}
}
```

https://github.com/golang/go/blob/39f2032c17516a55c165d329dd5e2e07f49132b0/src/io/fs/fs.go#L257

```go
func (e *PathError) Error() string { return e.Op + " " + e.Path + ": " + e.Err.Error() }
```